File: 64. Minimum Path Sum.py
# https://leetcode.com/problems/minimum-path-sum/

# Given a m x n grid filled with non-negative numbers, find a path from top left
# to bottom right, which minimizes the sum of all numbers along its path.

# Note: You can only move either down or right at any point in time.

################################################################################

# dp[m][n] = min_sum to (m,n)

class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        if not grid: return 0
        
        m, n = len(grid), len(grid[0])
        dp = [[0] * n for _ in range(m)]

        # base
        dp[0][0] = grid[0][0]
        for i in range(1, m): # first col
            dp[i][0] = dp[i-1][0] + grid[i][0]
        for j in range(1, n): # first row
            dp[0][j] = dp[0][j-1] + grid[0][j]

        # loop over row and col
        for i in range(1, m):
            for j in range(1, n):
                dp[i][j] = min(dp[i][j-1], dp[i-1][j]) + grid[i][j]
        
        return dp[-1][-1]

File: 64. Minimum Path Sum.cpp
// https://leetcode.com/problems/minimum-path-sum/

// Given a m x n grid filled with non-negative numbers, find a path from top left
// to bottom right, which minimizes the sum of all numbers along its path.

// Note: You can only move either down or right at any point in time.

////////////////////////////////////////////////////////////////////////////////

// dp[i][j] = minSum to (i,j)
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        return minPathSumDp1(grid);
        int m = grid.size(), n = grid[0].size();
        vector<vector<int>> dp(m, vector<int>(n));
        dp[0][0] = grid[0][0];
        for (int i = 1; i < m; ++i) {
            dp[i][0] = dp[i-1][0] + grid[i][0];
        }
        for (int j = 1; j < n; ++j) {
            dp[0][j] = dp[0][j-1] + grid[0][j];
        }
        for (int i = 1; i < m; ++i) {
            for (int j = 1; j < n; ++j) {
                dp[i][j] = min(dp[i][j-1], dp[i-1][j]) + grid[i][j];
            }
        }
        return dp.back().back();
    }
    int minPathSumDp1(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        vector<int> dp(n);
        dp[0] = grid[0][0];
        for (int j = 1; j < n; ++j) {
            dp[j] = dp[j-1] + grid[0][j];
        }
        for (int i = 1; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (j == 0) {
                    dp[j] += grid[i][j];
                } else {
                    dp[j] = min(dp[j-1], dp[j]) + grid[i][j];
                }
            }
        }
        return dp.back();
    }
};
